# As described in the README, my first attempt will be to try and classify this information according to distance/position. 

In [1]:
#import dependencies

import numpy as np
import pandas as pd

In [3]:
data = pd.read_csv('../data/training_set_metadata.csv')

Again further simplifying the question, I want to only answer this question based on objects captured with the same cadence. In this event, I am only working with "wide fast deep" images. (By confirming the ddf bool is false). 

In [9]:
wfd_data = data.loc[data['ddf'] == 0]

In [16]:
wfd_data

object_id          ra       decl       gal_l      gal_b  ddf  \
2116    1019335   85.605469 -30.691587  235.300564 -27.370560    0   
2117    1091307  145.019531  -9.291553  244.286275  31.057860    0   
2118    1107992  310.078125 -20.902014   24.531144 -32.885543    0   
2119    1135496   78.398438 -36.609226  240.359179 -34.593319    0   
2120    1135692  177.539062 -18.997131  283.433188  41.542691    0   
2121    1149951  302.871094 -16.646040   26.251596 -24.987488    0   
2122    1150736  158.027344 -16.334824  260.966474  34.945322    0   
2123    1153371  217.441406  -4.480799  343.587298  50.464355    0   
2124    1170601    4.042969  -5.379379  100.253364 -66.685650    0   
2125    1215304  290.039062 -15.868921   21.756580 -13.395677    0   
2126    1228479  356.132812 -11.872187   73.864458 -68.111084    0   
2127    1231302  344.355469  -1.492246   71.279268 -52.534180    0   
2128    1231534   44.472656  -9.291553  188.086335 -55.245046    0   
2129    1237057  181.933594   0.596842  279.553277  61.468883    0   
2130    1245166   81.792450 -60.623394  269.566561 -33.661441    0   
2131    1288208  172.720581 -52.029728  290.551110   8.894978    0   
2132    1306267  239.589844  -1.492246    8.178329  36.722884    0   
2133    1326217  354.726562  -2.537361   84.619419 -59.899405    0   
2134    1366732  193.183594 -20.742380  303.340623  42.128339    0   
2135    1408779  242.226562 -35.136116  342.350877  12.212392    0   
2136    1415987  124.628906 -18.209957  239.376145   9.848658    0   
2137    1426132   75.058594 -24.953020  225.761422 -34.660445    0   
2138    1454581  200.566406 -15.404094  313.821635  46.814793    0   
2139    1497514   58.004585 -49.312840  258.119980 -49.024729    0   
2140    1516088   51.503906 -13.861950  200.600617 -51.638868    0   
2141    1535056  311.484375 -10.655059   36.257649 -30.220400    0   
2142    1557442  328.383606 -46.571846  351.662172 -50.148858    0   
2143    1563124  248.378906 -22.669609  355.547663  16.808660    0   
2144    1573501  119.355469 -39.065495  254.550710  -5.229984    0   
2145    1594469  214.101562   1.044512  344.483305  56.897492    0   
...         ...         ...        ...         ...        ...  ...   
7818  130219752  355.163544 -50.091457  328.699808 -63.346833    0   
7819  130231675   22.851562 -18.681826  172.442571 -77.518364    0   
7820  130263372   86.132812 -13.708317  218.054844 -20.910333    0   
7821  130312781  133.242188 -24.788561  249.573365  12.560887    0   
7822  130319749   11.074219  -8.084014  117.528357 -70.880802    0   
7823  130330088  340.839844 -32.442867   14.265035 -61.816065    0   
7824  130359176  138.164062 -40.033035  264.026219   5.768998    0   
7825  130375489  356.132812 -27.111860   29.160507 -75.123425    0   
7826  130386135   28.828125 -25.117701  208.535318 -75.532469    0   
7827  130402542  140.976562   3.732834  228.783975  35.301875    0   
7828  130408188  314.296875 -40.033035    2.071417 -40.445067    0   
7829  130414189   36.480583 -51.642994  273.384272 -59.721065    0   
7830  130489916  209.707031 -38.300922  317.204568  22.695931    0   
7831  130552230   54.316406 -18.524391  209.170030 -51.015495    0   
7832  130595291  304.980469 -14.169522   29.658085 -25.884722    0   
7833  130617044   68.378906 -16.334824  213.119133 -37.716403    0   
7834  130622528  132.539062  -1.641510  229.063354  25.304821    0   
7835  130639669  164.003906 -31.039240  275.630181  25.642742    0   
7836  130659834  174.902344  -2.089372  269.456940  56.041993    0   
7837  130678775  142.734375 -23.480536  254.430952  19.977762    0   
7838  130684460  117.949219  -0.895283  220.752608  12.975772    0   
7839  130695262   90.000000 -20.264481  226.045011 -20.105392    0   
7840  130698059  142.734375 -18.997131  250.892051  23.021357    0   
7841  130716752  337.372894 -57.209084  332.816119 -50.876643    0   
7842  130727624  294.960938  -4.031936   34.856817 -12.602145    0   
7843  130739978 

Next, galactic coordinates are a better indicator of position when considering far off astronomical phenomena, not to mention is also ra/dec provide position, so it is redundant to keep both. I will now remove one coord. system. 

In [17]:
del wfd_data['ra'], wfd_data['decl']

In [21]:
wfd_data['target'].unique()

array([90, 52, 42, 15, 65, 16, 95,  6, 67, 64, 62, 88, 92, 53])

In [23]:
wfd_data = wfd_data[['gal_l','gal_b','hostgal_specz','target']]

In [25]:
wfd_data.dropna(inplace=True)

/Users/noahkasmanoff/anaconda/lib/python3.6/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


In [27]:
wfd_data

gal_l      gal_b  hostgal_specz  target
2116  235.300564 -27.370560         0.8621      90
2117  244.286275  31.057860         0.3297      52
2118   24.531144 -32.885543         0.1426      42
2119  240.359179 -34.593319         0.2460      90
2120  283.433188  41.542691         0.4408      15
2121   26.251596 -24.987488         0.3539      90
2122  260.966474  34.945322         0.0950      42
2123  343.587298  50.464355         0.0000      65
2124  100.253364 -66.685650         0.5727      90
2125   21.756580 -13.395677         0.0000      16
2126   73.864458 -68.111084         0.3173      90
2127   71.279268 -52.534180         1.8247      95
2128  188.086335 -55.245046         0.1223      42
2129  279.553277  61.468883         0.1784      42
2130  269.566561 -33.661441         0.3033      90
2131  290.551110   8.894978         0.0000      16
2132    8.178329  36.722884         0.0682      42
2133   84.619419 -59.899405         0.4106      90
2134  303.340623  42.128339         0.1174      42
2135  342.350877  12.212392         0.3237      90
2136  239.376145   9.848658         0.0000      16
2137  225.761422 -34.660445         0.3796      90
2138  313.821635  46.814793         0.1172      15
2139  258.119980 -49.024729         0.0000       6
2140  200.600617 -51.638868         0.0000       6
2141   36.257649 -30.220400         0.1159      15
2142  351.662172 -50.148858         0.2231      67
2143  355.547663  16.808660         0.1614      15
2144  254.550710  -5.229984         0.0355      64
2145  344.483305  56.897492         0.1066      42
...          ...        ...            ...     ...
7818  328.699808 -63.346833         0.0000       6
7819  172.442571 -77.518364         0.0949      42
7820  218.054844 -20.910333         0.2456      90
7821  249.573365  12.560887         0.0000      16
7822  117.528357 -70.880802         0.0760      42
7823   14.265035 -61.816065         0.0000      65
7824  264.026219   5.768998         0.0397      52
7825   29.160507 -75.123425         0.2886      90
7826  208.535318 -75.532469         0.0000      65
7827  228.783975  35.301875         0.1518      15
7828    2.071417 -40.445067         0.0000      16
7829  273.384272 -59.721065         0.0542      90
7830  317.204568  22.695931         0.0000      16
7831  209.170030 -51.015495         0.1226      42
7832   29.658085 -25.884722         0.2094      15
7833  213.119133 -37.716403         0.1423      52
7834  229.063354  25.304821         0.0000      65
7835  275.630181  25.642742         0.2760      62
7836  269.456940  56.041993         0.4215      42
7837  254.430952  19.977762         0.0584      64
7838  220.752608  12.975772         0.0000      16
7839  226.045011 -20.105392         0.1273      15
7840  250.892051  23.021357         0.0591      67
7841  332.816119 -50.876643         0.0000      16
7842   34.856817 -12.602145         0.0000      65
7843  172.342697 -72.255675         0.0000      65
7844  275.742955 -16.509746         0.1725      90
7845  308.728904   6.727511         0.0000      16
7846  239.172243 -33.827844         0.0000      65
7847   25.102988 -24.511101         0.0000       6

[5732 rows x 4 columns]

Now to make a rough model. Can you determine the type of source based on the position and redshift of that object? Let's find out.

In [31]:
y = wfd_data['target']

y = pd.get_dummies(y)

del wfd_data['target']

In [34]:
X = wfd_data.values
y = y.values

In [35]:
from sklearn.model_selection import train_test_split


In [36]:
X_train,X_test,y_train,y_test = train_test_split(X,y)

In [38]:
len(X_test)

1433

In [43]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.multioutput import MultiOutputClassifier


In [44]:
rf = MultiOutputClassifier(RandomForestClassifier())

In [45]:
rf.fit(X_train,y_train)

MultiOutputClassifier(estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False),
           n_jobs=1)

In [46]:
rf.score(X_test,y_test)

0.209351011863224

# Outcome: Based on just location and distance, you get a testing accuracy of ~21%, which is better than random ( 1/ 14 shot = 7%), but still terrible. Furthermore there is supposedly a 15th class to consider, and that has not been included here. So in conclusion, this method fails. 